plugin/trino-redshift/src/main/java/io/trino/plugin/redshift/RedshiftUnloadSplitManager.java

```diff
@@ -0,0 +1,195 @@
+        }
+        JdbcTableHandle jdbcTableHandle = dynamicFilteringEnabled(session) ? ((JdbcTableHandle) table).intersectedWithConstraint(dynamicFilter.getCurrentPredicate()) : (JdbcTableHandle) table;
+        List<JdbcColumnHandle> columns = jdbcTableHandle.getColumns()
+                .orElseGet(() -> jdbcClient.getColumns(
+                        session,
+                        jdbcTableHandle.getRequiredNamedRelation().getSchemaTableName(),
+                        jdbcTableHandle.getRequiredNamedRelation().getRemoteTableName()));
+
+        if (!isUnloadSupported(((JdbcTableHandle) table), columns)) {
+            log.debug("Unload query contains unsupported characters. Falling back to using JDBC");
+            return fallbackSplitSource;
+        }
+
+        Connection connection;
+        PreparedStatement statement;
+        try {
+            connection = jdbcClient.getConnection(session);
+            String redshiftSelectSql = getRedshiftSelectSql(session, connection, jdbcTableHandle, columns);
+            if (redshiftSelectSql.contains("\\b")) {
+                log.debug("Unload query contains unsupported characters. Falling back to using JDBC");
+                return fallbackSplitSource;
+            }
+            statement = buildUnloadSql(session, connection, columns, redshiftSelectSql);
+        }
+        catch (SQLException e) {
+            if (e instanceof RedshiftException) {
+                log.debug("Could not build unload query. Falling back to using JDBC");
+                return fallbackSplitSource;
+            }
+            throw new RuntimeException(e);
+        }
+        return new RedshiftUnloadSplitSource(executor, connection, statement, fallbackSplitSource);
+    }
+
+    private String getRedshiftSelectSql(ConnectorSession session, Connection connection, JdbcTableHandle table, List<JdbcColumnHandle> columns)
+            throws SQLException
+    {
+        PreparedQuery preparedQuery = jdbcClient.prepareQuery(session, table, Optional.empty(), columns, ImmutableMap.of());
+        PreparedStatement openTelemetryPreparedStatement = queryBuilder.prepareStatement(jdbcClient, session, connection, preparedQuery, Optional.of(columns.size()));
+        RedshiftPreparedStatement redshiftPreparedStatement = openTelemetryPreparedStatement.unwrap(RedshiftPreparedStatement.class);
+        String selectQuerySql = redshiftPreparedStatement.toString();
+        return queryModifier.apply(session, selectQuerySql); // TODO is this required?
+    }
+
+    private PreparedStatement buildUnloadSql(ConnectorSession session, Connection connection, List<JdbcColumnHandle> columns, String redshiftSelectSql)
+            throws SQLException
+    {
+        String unloadSql = "UNLOAD ('%s') TO '%s' %s FORMAT PARQUET".formatted(
+                formatStringLiteral(redshiftSelectSql), // TODO does it require any other escaping
+                unloadLocation + session.getQueryId() + "-" + UUID.randomUUID() + "/",
```

Comment:
why do we have the uuid at the end?

do we have any mechanism to figure out from the client perspective the location of the unloaded content?

Reply:
ya, I don't think we need uuid, just queryid should be enough

Reply:
A single Trino query can load multiple fragments from the same connector, all of them must use a unique location.

Reply:
ah, right